### ISSUE TITLE
Set up dedicated devices

### ISSUE THREAD

**Issue by debian-user-france1** [marcelklehr/recognize#73]
Make an app that can be installed on other devices and a setting that if a device that has such app and is configured (and online) gets the model & pictures so it can process byself. My Server doesn't have a good GPU so I would like to run it on my fast er computer with a NVIDIA GeForce GTX 1660 super (or if a rasperry pi is online send it some data so it is a bit faster etc.)

**Comment by marcelklehr** [comment]
This is a good idea, but it will take some time until I can get to it.

**Comment by debian-user-france1** [comment]
I love this project

**Comment by stavros-k** [comment]
One advantage is that you don't have to overload a nextcloud container with extra packages.
Also it can be scaled up a lot, like in a kubernetes cluster, you can spawn multiple workers to do the processing. (maybe not so much for a home lab, but in a bigger installations would be awesome).

**Comment by debian-user-france1** [comment]
Maybe I can do a little beta

Maybe a bit of bash script that lets the classifier work

**Comment by UmbrellaCodr** [comment]
+1, as I don't have support for glibc, having another docker image with all the supported libraries makes a lot more sense.

**Comment by gotagitaccnt** [comment]
Any updates on this? Would really appreciate this feature to be implemented or maybe some pointers how to start

**Comment by guystreeter** [comment]
I can run the rest of Nextcloud on a 1GB VM.
A 4GB VM costs almost 4 times a much per month.
I would like to to batch-process recognition on another machine that I only spin up for a day once in a while. (Or do the recognition runs on my home machine with its NVIDIA GPU)

**Comment by ddarek2000** [comment]
Great Idea.  Up-Vote

**Comment by szaimen** [comment]
I think this would also be a good idea e.g. for AIO since libtensorflow does not seem to run in an alpine container but could then run from another container which could use debian as base.

**Comment by guystreeter** [comment]
Setting up GPU access for a container is complicated. Anyone running Nextcloud in a container would probably want to send image analysis requests to a service running on the native OS.

**Comment by marcelklehr** [comment]
> Setting up GPU access for a container is complicated.

It's not that hard, I believe

**Comment by guystreeter** [comment]
> > Setting up GPU access for a container is complicated.
> 
> It's not that hard, I believe

Have you tried it? The documented steps are expert admin level stuff

**Comment by relink2013** [comment]
I would love for this to become a reality. Im currently running NC in a Ubuntu VM for the sole purpose of using my Nvidia GPU with Recognize and Memories and I absolutely hate managing it.

**Comment by tbelway** [comment]
> > > Setting up GPU access for a container is complicated.
> > 
> > 
> > It's not that hard, I believe
> 
> Have you tried it? The documented steps are expert admin level stuff

I don't know what you mean it's expert admin level stuff. It's pretty simple... 
I've done this both through debian and RHEL based hypervisors (proxmox and kvm with cockpit) via both direct lxc or podman containerization as well as through hardware passthrough to a VM which is then containerized. As long as you are running linux, it's trivial.

**Comment by Leptopoda** [comment]
@marcelklehr 
I currently have some time to look into this and would love giving this a shot.
What is the blocker? Or how would you imagine this being implemented?

**Comment by arch-user-france1** [comment]
> @marcelklehr I currently have some time to look into this and would love giving this a shot. What is the blocker? Or how would you imagine this being implemented?

IMO, a nodejs socket server could be run on the dedicated device and then the images sent and tagged with an ID; preferably 128 to 512 at once, because that's the amount modern graphics cards handle with 100% utilisation and not something like 7% because the images are not supplied in-time or do not have enough pixels.
The dedicated device would send the result back with the appropriate ID, and done.

To set up a socket, `socket.io` or `ws` could be used.

It would also be very handy to have a configuration file. The easiest would be to write the variables exported into a javascript file and then import them in the actual program. Better, though, would be CSV files or something like them.


Remember tis a simple draft. Per'aps someone would like to implement it. There are many solutions to the problem, and it would always be great if we could get an anwer from _marcelklehr_ so we know what he actually likes to have.

**Comment by Leptopoda** [comment]
Why would a nodejs socket server be needed?
Tensorflow serving already has a restfull api that we could interact with directly. Also the latest version introduces a batch size option that sound like what you wanted.

I think you mean the training device (running TFserving) should fetch the jobs from the server but that's not how TFserving is meant to be used.

**Comment by arch-user-france1** [comment]
I meant not to use TFserving, but make it on my own. But in case you want to use it, yes, may be better.

**Comment by 617a7a** [comment]
Any update on this?

**Comment by Tsaukpaetra** [comment]
(Commenting to add myself to notifications)
I am also interested in testing anything that can help this proceed. I recently updated from NC 23 (phew) and saw this nifty app. I enabled it, and then made a 😓 face when I realized the dinky little Celeron the server was running under would take years to process the current files (let alone any new ones that will ingest as part of the family archive project). Meanwhile my gaming PC is sitting idle and I'm left pondering how to bridge the gap of immense power for the times it is needed on demand. :)

**Comment by RudolfAchter** [comment]
Same situation here. Looks to me like a huge pain to get tensorflow running in the nextloud-aio alpine image. For me there also would be the benefit of using the power of my gamin pc.
Thinking of enterprises you could offload the gpu load to different GPU Servers or even a GPU Server cluster for analyzing stuff.

**Comment by marcelklehr** [comment]
Nextcloud GmbH is planning to move the classifiers in recognize to docker containers as part of the External Apps Ecosystem in the coming months

**Comment by szaimen** [comment]
> Nextcloud GmbH is planning to move the classifiers in recognize to docker containers as part of the External Apps Ecosystem in the coming months

Sounds great! As soon as it is available via the External Apps Ecosystem, it will also automatically be available in AIO after one enables the docker socket proxy in the AIO interface :)

**Comment by marcelklehr** [comment]
> Nextcloud GmbH is planning to move the classifiers in recognize to docker containers as part of the External Apps Ecosystem in the coming months

Sorry to say, the plans have been scrapped due to lack of engineering time so far. It's still on our list of things that would be nice to have, but it's not scheduled any time soon for now :/

As mentioned in #1061 I'd be open to community contributions on this.

My rough plan would be not to deviate too much from how the models are run right now. Instead of the [Classifier class executing node.js directly](https://github.com/nextcloud/recognize/blob/main/lib/Classifiers/Classifier.php?rgh-link-date=2024-04-18T11%3A02%3A14Z#L165), there would be an option in the settings to call out to the recognize External App instead, or perhaps the external app could be auto-detected. The external app would do the same thing as the Classifier class, execute node.js and return the json line results, so they can be processed in the original recognize app. These are the current docs on how App API / External Apps work: [cloud-py-api.github.io/app_api/index.html](https://cloud-py-api.github.io/app_api/index.html)

**Comment by marcelklehr** [comment]
Note, that you do have a little influence over what Nextcloud GmbH works on: We don't promise anything, but every release cycle we try to work on enhancements that get a lot of upvotes, so you may express your support for this by giving this issue an upvote.

**Comment by bugsyb** [comment]
Upvote for this as it's required for https://github.com/nextcloud/recognize/issues/69.

**Comment by marcelklehr** [comment]
By upvote, I mean, adding a thumbs-up reaction on the original post ;)

**Comment by priyankub** [comment]
Why do we not have a debian/non-alpine image based AIO container for more capable systems?

**Comment by Zoey2936** [comment]
> Why do we not have a debian/non-alpine image based AIO container for more capable systems?
> 
> I launched one locally for my system, but before I spend more time, want to check if there is even an appetite to launch something for users with more capable systems

see https://github.com/nextcloud/all-in-one/discussions/3382

**Comment by priyankub** [comment]
Thank you! So likely there's no appetite to move to Debian based images anytime soon.
As such, expanding recognize to be able to use external container seems to be the only other viable option.

Did I understand it right?

**Comment by priyankub** [comment]
What base image would the team prefer for a container that can use GPU to run the classifiers?

**Comment by marcelklehr** [comment]
We've been using nvidia/cuda:12.2.2-cudnn8-devel-ubuntu22.04 for our AI ex apps so far.

**Comment by ngrigoriev** [comment]
Upvoting for the following reason: I run Nextcloud on a small ARM SBC and it is sufficient for most of the needs. The heavy compute needed to process the images would be idling 99.9% of the time. I would love to run the indexing on my powerful desktop once in a while to reindex the pictures stored on the NAS anyway. This sort of functionality is good for the cloud infrastructure where you can have compute on-demand, but for self-hosting environment, especially the small home one, you usually have some kind of low-power server(s), NAS(es) and 100x more powerful desktop uses occasionally :)

**Comment by priyankub** [comment]
> What base image would the team prefer for a container that can use GPU to run the classifiers?
> 
> Context - I am currently working and testing out this feature

I am no longer motivated to work on this, and hope someone else takes it up for community contribution!

After having indexed 250k+ files in my server, newer files that are added would be < 1000/month and will be CPU indexed pretty quickly

**Comment by BjoKaSH** [comment]
> > What base image would the team prefer for a container that can use GPU to run the classifiers?
> > Context - I am currently working and testing out this feature
> 
> I am no longer motivated to work on this, and hope someone else takes it up for community contribution!
> 
> After having indexed 250k+ files in my server, newer files that are added would be < 1000/month and will be CPU indexed pretty quickly

Hi @priyankub ,  fully understood that you are no longer interested in this work, needs simply change over time, and great to read you got your needs covered.

However, would you mind to share what ever small or large progress you made?  Any implementation or architecture work you did might help others to pick up the work.

**Comment by lucasew** [comment]
> I can run the rest of Nextcloud on a 1GB VM. A 4GB VM costs almost 4 times a much per month. I would like to to batch-process recognition on another machine that I only spin up for a day once in a while. (Or do the recognition runs on my home machine with its NVIDIA GPU)

What I plan doing here is that I have a Hostinger VM that is 2vCPU 8GB RAM which runs nextcloud, and other stuff

But I have my RTX 3060 on the same "network" via Tailscale, I start the inference container in this PC, find a way to start that inference container, trigger a job to run inferences on the queue then stop the inference container.

I rarely am awake between 2am and 6am so it would be the perfect window lol.

**Comment by nismanoku** [comment]
Would love to see an ExApps for recognize with support for intel (igpu/arc). If i can help with testing

**Comment by lord0gnome** [comment]
This issue interests me, and I'd be ready to try and tackle it, with some support from people who know how recognize itself works. If someone has a discord or something of the likes to posit questions I am interested.

**Comment by marcelklehr** [comment]
@lord0gnome I sent you a mail :)

**Comment by szaimen** [comment]
🥳🥳🥳🥳🥳

**Comment by marcelklehr** [comment]
v13.0.0 will soon be published for the upcoming nc 35 release, which integrates with the recognize_backend exApp.
